### PR TITLE
[expo-notifications] Add clearLastNotificationResponseAsync to API

### LINKED ISSUE #18712: useLastNotificationResponse clears only when app is killed.
### Summary

We are using the `useLastNotificationResponse` hook to check against tapped notifications and route the user to a specific screen. 

**However, we can't figure out how to clear the data from that hook at all. The last notification data only clears out when the app is killed.**

In our case, our users are now always being routed until they kill the app. 

We can figure out a hacky way where we save the last notification and do comparisons against incoming ones but that's obviously not ideal.

### What platform(s) does this occur on?

Android, iOS

### Environment

```
  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 12.2.1
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 16.15.0 - /usr/local/bin/node
      Yarn: 1.22.19 - ~/Documents/work/mobile/node_modules/.bin/yarn
      npm: 8.5.5 - /usr/local/bin/npm
    Managers:
      CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
    IDEs:
      Xcode: 13.4/13F17a - /usr/bin/xcodebuild
    npmPackages:
      expo: ^45.0.0 => 45.0.6 
      react: 17.0.2 => 17.0.2 
      react-dom: 17.0.2 => 17.0.2 
      react-native: 0.68.2 => 0.68.2 
      react-native-web: 0.17.7 => 0.17.7 
    npmGlobalPackages:
      eas-cli: 0.53.1
      expo-cli: 6.0.2
    Expo Workflow: managed
```

### Minimal reproducible example

```
const lastNotification = Notifications.useLastNotificationResponse();

const handleBackgroundNotification = async () => {
    const route = lastNotification?.notification.request.content.data.route;			
    routeUser(data);
    await clearNotifications();
};

useEffect(() => {
    handleBackgroundNotification();
}, [lastNotification]);
```

### PR DESCRIPTION
**Pull request by douglowder**
# Why

A method is needed to clear the last notification response from memory.

Fixes https://github.com/expo/expo/issues/18712

# How

Add new API to the NotificationEmitter module.

# Test Plan

- CI should pass
- New API should work as expected in the test app.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by douglowder** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/expo/expo/31607?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#31607** <a href="https://app.graphite.dev/github/pr/expo/expo/31607?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#31598** <a href="https://app.graphite.dev/github/pr/expo/expo/31598?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @douglowder and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/expo-notifications/src/NotificationsEmitter.ts [hunk 1]
```diff
@@ -109,9 +109,13 @@ export function removeNotificationSubscription(subscription: EventSubscription)
...
 
-// @docsMissing
 /**
- * @header listen
+ * Gets the notification response that was received most recently
+ * (a notification response designates an interaction with a notification, such as tapping on it).
+ *
+ * @return A promise that resolves to one of these values:
+ * - `null` - if no notification response has been received yet,
+ * - a [`NotificationResponse`](#notificationresponse) object - if a notification response was received.
```
Comment (amandeepmittal) [inline packages/expo-notifications/src/NotificationsEmitter.ts:118]:
```suggestion
 * - `null` - if no notification response has been received yet
 * - a [`NotificationResponse`](#notificationresponse) object - if a notification response was received
```

File: packages/expo-notifications/src/NotificationsEmitter.ts [hunk 2]
```diff
@@ -121,3 +125,17 @@ export async function getLastNotificationResponseAsync(): Promise<NotificationRe
   const mappedResponse = response ? mapNotificationResponse(response) : response;
   return mappedResponse;
 }
+
+/**
+ * Clears the notification response that was received most recently. May be used in cases
+ * where the app selects a route based on the notification response, and it is undesirable
+ * to continue to select the route after the response has already been handled.
```
Comment (amandeepmittal) [inline packages/expo-notifications/src/NotificationsEmitter.ts:132]:
```suggestion
 * Clears the notification response that was received most recently. May be used 
 * when an app selects a route based on the notification response, and it is undesirable
 * to continue selecting the route after the response has already been handled.
```

A suggestion to rephrase the second sentence to improve readability.

File: packages/expo-notifications/src/NotificationsEmitter.ts [hunk 3]
```diff
@@ -121,3 +125,17 @@ export async function getLastNotificationResponseAsync(): Promise<NotificationRe
...
+ *
+ * @return A promise that resolves if the native call was succesful.
```
Comment (vonovak) [inline packages/expo-notifications/src/NotificationsEmitter.ts:134]: typo: succesful -> successful

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/emitting/NotificationsEmitter.kt [hunk 4]
```diff
@@ -41,6 +41,11 @@ open class NotificationsEmitter : Module(), NotificationListener {
     AsyncFunction<Bundle?>("getLastNotificationResponseAsync") {
       lastNotificationResponseBundle
     }
+
+    AsyncFunction("clearLastNotificationResponseAsync") {
+      lastNotificationResponseBundle = null
```
Comment (vonovak) [inline packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/emitting/NotificationsEmitter.kt:46]: If I'm not mistaken, this clears `lastNotificationResponseBundle` but the `useLastNotificationResponse` hook which is mentioned in the original issue https://github.com/expo/expo/issues/18712 will still have the old state?

Comment (douglowder) [reply packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/emitting/NotificationsEmitter.kt:46]: I'll have a look to see if I can reproduce this and come up with a clean fix.

Comment (douglowder) [reply packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/emitting/NotificationsEmitter.kt:46]: Fixed!

File: packages/expo-notifications/src/NotificationsEmitter.ts [hunk 5]
```diff
@@ -121,3 +126,30 @@ export async function getLastNotificationResponseAsync(): Promise<NotificationRe
...
+  return await NotificationsEmitterModule.clearLastNotificationResponseAsync();
+}
+
+/**
+ * @hidden
+ */
+export function addNotificationResponseClearedListener(listener: () => void): EventSubscription {
+  return emitter.addListener<void>(didClearNotificationResponseEventName, () => {
+    listener();
+  });
```
Comment (vonovak) [inline packages/expo-notifications/src/NotificationsEmitter.ts:154]:
```suggestion
  return emitter.addListener<void>(didClearNotificationResponseEventName, listener);
```

Not a big difference in this case (saving the function allocation), but I find this equally readable.

**Review by vonovak — APPROVED** [review]
LGTM, though I wonder if the event could be emitted directly from the native module, as opposed to changing state in native land and emitting the corresponding event from JS (+ the event is emitted prior to the native change taking place). It'd feel a little nicer to me. (probably a subjective feeling :D)

**Review by tsapeta — APPROVED** [review]

File: packages/expo-notifications/src/NotificationsEmitter.ts [hunk 6]
```diff
@@ -121,3 +126,30 @@ export async function getLastNotificationResponseAsync(): Promise<NotificationRe
...
+
+/**
+ * Clears the notification response that was received most recently. May be used in cases
+ * where the app selects a route based on the notification response, and it is undesirable
+ * to continue to select the route after the response has already been handled.
+ *
+ * If a component is using the [`useLastNotificationResponse`](#useLastNotificationResponse) hook,
+ * this call will also clear the value returned by the hook.
+ *
+ * @return A promise that resolves if the native call was succesful.
```
Comment (douglowder) [inline packages/expo-notifications/src/NotificationsEmitter.ts:138]:
```suggestion
 * @return A promise that resolves if the native call was successful.
```

**Comment by avishekdev** [comment]
@douglowder is this fix live now? If yes then to which version I have to bump my expo-notification package? Thanks

**Comment by douglowder** [comment]
This is not published yet. I'll discuss with the rest of the team about whether this can be cherry picked for the SDK 51 maintenance branch.

**Comment by avishekdev** [comment]
Cool thanks man @douglowder. You guys are doing a great job.🙏

> This is not published yet. I'll discuss with the rest of the team about whether this can be cherry picked for the SDK 51 maintenance branch.

**Comment by Vumpik** [comment]
Hey @douglowder any chance that it could go live in any following days? Or do you know if there is any workaround that I could integrate? Only one that I can think of is to save ID and check if those ID was already shown/processed and ignore if yes.


We would like to go to the production. Thank you really much for the fix!

**Comment by vonovak** [comment]
Hi! We will release this in a few days but if you need the changes now you can use a tool like patch-package to apply the changes locally.

**Comment by Vumpik** [comment]
Yes. I know about that. But wanted to go with a cleaner-ish solution :) Thank you so much! :)

> Hi! We will release this in a few days but if you need the changes now you can use a tool like patch-package to apply the changes locally.

**Comment by douglowder** [comment]
@Vumpik this is published in `expo-notifications@0.28.18`

**Comment by avishekdev** [comment]
@douglowder, should I bump up the expo-notifications version now I am using `"expo": "~50.0.14",`?

**Comment by douglowder** [comment]
> @douglowder, should I bump up the expo-notifications version now I am using `"expo": "~50.0.14",`?

@avishekdev the new version of expo-notifications is an SDK 51 package. You should do

```
npx expo install expo@^51.0.33 --fix
npx expo install expo-notifications@^0.28.18
```

**Comment by justinkx** [comment]
@douglowder Issue still persist on expo@51.0.38 and expo-notifications@0.28.18.
After opening a Push notification using **getLastNotificationResponseAsync** and later if app is restarted using **Updates.reloadAsync()** , getLastNotificationResponseAsync returns data even though it was cleared using **clearLastNotificationResponseAsync**

```javascript
async function readLastNotificationAsync() {
  try {
    const request = await Notifications.getLastNotificationResponseAsync()
    if (request) {
      await redirectToScreen(request?.notification)
      Notifications.clearLastNotificationResponseAsync()
    }
  } catch (error) {}
}
```

**Comment by justinkx** [comment]
> @douglowder Issue still persist on expo@51.0.38 and expo-notifications@0.28.18. After opening a Push notification using **getLastNotificationResponseAsync** and later if app is restarted using **Updates.reloadAsync()** , getLastNotificationResponseAsync returns data even though it was cleared using **clearLastNotificationResponseAsync**
> 
> ```js
> async function readLastNotificationAsync() {
>   try {
>     const request = await Notifications.getLastNotificationResponseAsync()
>     if (request) {
>       await redirectToScreen(request?.notification)
>       Notifications.clearLastNotificationResponseAsync()
>     }
>   } catch (error) {}
> }
> ```

Further Debugging found that upon restart of app using **Updates.reloadAsync()** , I am getting log `NotificationsEmitter.onNotificationResponseIntentReceived` at [NotificationsEmitter onNotificationResponseIntentReceived](https://github.com/expo/expo/blob/main/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/emitting/NotificationsEmitter.kt#L80C29-L80C86). Can this be due to [ExpoNotificationLifecycleListener.java](https://github.com/expo/expo/blob/main/packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/ExpoNotificationLifecycleListener.java#L41)

**Comment by justinkx** [comment]
> > @douglowder Issue still persist on expo@51.0.38 and expo-notifications@0.28.18. After opening a Push notification using **getLastNotificationResponseAsync** and later if app is restarted using **Updates.reloadAsync()** , getLastNotificationResponseAsync returns data even though it was cleared using **clearLastNotificationResponseAsync**
> > ```js
> > async function readLastNotificationAsync() {
> >   try {
> >     const request = await Notifications.getLastNotificationResponseAsync()
> >     if (request) {
> >       await redirectToScreen(request?.notification)
> >       Notifications.clearLastNotificationResponseAsync()
> >     }
> >   } catch (error) {}
> > }
> > ```
> 
> Further Debugging found that upon restart of app using **Updates.reloadAsync()** , I am getting log `NotificationsEmitter.onNotificationResponseIntentReceived` at [NotificationsEmitter onNotificationResponseIntentReceived](https://github.com/expo/expo/blob/main/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/emitting/NotificationsEmitter.kt#L80C29-L80C86). Can this be due to [mPendingNotificationResponsesFromExtras](https://github.com/expo/expo/blob/main/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/NotificationManager.java#L64) still having Bundle preserved?

@douglowder Tried this patch and it worked
```patch
diff --git a/node_modules/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/NotificationManager.java b/node_modules/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/NotificationManager.java
index 867c70d..436f86f 100644
--- a/node_modules/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/NotificationManager.java
+++ b/node_modules/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/NotificationManager.java
@@ -74,6 +74,7 @@ public class NotificationManager implements SingletonModule, expo.modules.notifi
   @Override
   public void removeListener(NotificationListener listener) {
     mListenerReferenceMap.remove(listener);
+    mPendingNotificationResponsesFromExtras.clear();
   }
 
   /**
diff --git a/node_modules/expo-notifications/ios/EXNotifications/Notifications/Emitter/EXNotificationsEmitter.m b/node_modules/expo-notifications/ios/EXNotifications/Notifications/Emitter/EXNotificationsEmitter.m
index 7d46d62..f692ac9 100644
--- a/node_modules/expo-notifications/ios/EXNotifications/Notifications/Emitter/EXNotificationsEmitter.m
+++ b/node_modules/expo-notifications/ios/EXNotifications/Notifications/Emitter/EXNotificationsEmitter.m
@@ -83,8 +83,7 @@ - (void)application:(UIApplication *)application didReceiveRemoteNotification:(N
 
 - (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler
 {
-  _notificationCenterDelegate.lastNotificationResponse = response;
-  [self sendEventWithName:onDidReceiveNotificationResponse body:[self serializedNotificationResponse:response]];
+  [self sendEventWithName:onDidReceiveNotificationResponse body:[self serializedNotificationResponse:_notificationCenterDelegate.lastNotificationResponse]];
   completionHandler();
 }
```

**Pull request merged**